src/sonic-ctrmgrd/ctrmgr/ctrmgrd.py

```diff
@@ -426,6 +430,40 @@ def do_join(self, ip, port, insecure):
                     format(remote_ctr_config[JOIN_RETRY], self.start_time))
 
 
+def tag_latest_image(server, feat, docker_id, image_ver):
+    res = 1
+    if not UNIT_TESTING:
+        status = os.system("docker ps |grep {} >/dev/null".format(docker_id))
```

Comment:
Currently, I set "tag_latest_image_on_wait_seconds" = 600s, it means when a k8s pod running for over 10 minutes(by monitor the pod's container id), I will tag the container's image. About this default value, do you have any ideas?

Reply:
May be not. Suggest
1) Try restarting service during this period
2) Kill the main process in the container that will make the container exit within this period
3) Any more .... ;)

Reply:
The situation you say should be:
- k8s schedules container and remote_state is running
- ctrmgrd capture remote_state ready -> running during last process, register timer_handler(600s, tag_function(previous_docker_id))
- during the 600s, systemctl restart feature
The current reactions are:
- feature systemd service will stop -> start -> wait, when stop, set remote_state=stopped, when start, not docker start, when wait, wait docker_id
- k8s finds its scheduled container is killed, try to reschedule, container_up set remote_state=pending
- ctrmgrd captures remote_state=pending, set remote_state=ready
- container_up set remote_state=running and set docker_id
- ctrmgrd capture remote_state ready -> running, will register timer_handler(600s, tag_function(current_docker_id))
- and tag_latest_image_on_wait_seconds is up, tag function finds the previous_docker_id is disappeared, will do nothing
- and tag_latest_image_on_wait_seconds is up, tag function finds the current_docker_id, tag latest.
Have tested this case and some other cases, all good.
Any other corner cases you want me to test?

And by the way, about removing 'tag as local' from reboot script, you want me to do it in the PR or in the future? I want to do in the future, because it seems redundant, but no real impacts. And it can ensure tag as local when reboot, maybe necessary for us.

Reply:
I prefer we do it in this PR, as this PR's major purpose is to tag local.

One more Q: After you tag as "local" when you do service restart, who wins, systemd or kube. 

I prefer systemd takes over. This would add the new version in label that should auto-block kube from deploying. 
Can you check this out too?

Reply:
Tag as local is in reboot script in sonic-utilities repo, removing from reboot script PR is here. [#](https://github.com/sonic-net/sonic-utilities/pull/2451)

Q: After you tag as "local" when do service restart, who wins?
    Kube wins.
    I see label such as feature_version_enabled=false will be set only when containers start up as local. Actually, when we restart the service, systemd set remote_state=stopped so that systemd will wait k8s to reschedule. K8s finds its container is down, will reschedule immediately. In this process, container has no change to start up as local to set feature_version_enabled=false. So, there is no auto-block.
About this problem, I think it's right that service is taken over by kube in this situation, because the set-owner is kube and k8s is connected. Taken over by kube should be the proper action.


By the way. I want to remove the instance_higher checker, because we want to support fallback in the future, if post-check failed, we could fall back to the previous version. What's your concern about this before, is there any necessary things so that we always need the higher version?

Reply:
Two points:
1) Please merge the reboot fix PR into this PR -- Like to keep all related in one PR. 
    For example, if we revert one, it should affect both. it should both either stay or go as one piece.
2) I agree with your explanation. Earlier tagging upon reboot only did not need any change, as state-db starts empty. Now that we are preponing the tagging, we need to do required updates that systemd takes over, if kube image is tagged as latest. In fact, if it makes simple, we can change from "kube running" to  "kube-transient" once you tag it latest. On next systemd restart, seeing it kube-transient" we can go local path.

No need for kube to manage after local tagging. It is more stable if systemd takes over.

Reply:
Point 1: About the two PRs, the reason is that the reboot script is in another repo called sonic-utilities, seems we can't raise a PR including two repos code.

Point 2: About after tagging, it's easy to let feature go back to local after restart, but one of our visions is that k8s can collect the containers real-time metrics, if features are taken over by systemd after tagging, we can't keep collecting the metrics.

Reply:
Point 1 -- Agree.
Point 2 -- Debatable. What happens if device gets rebooted ? We have STATE-DB for counters/metrics.

Let it be so for now.